### PR TITLE
CC: Remote hypervisor merge to main

### PR DESCRIPTION
**Pull request by stevenhorsman**
Forward port remote-hypervisor support from `CCv0` 

Fixes: #6867

### CODE REVIEW

File: src/runtime/virtcontainers/remote.go [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+	annotations[hypannotations.DefaultVCPUs] = strconv.FormatUint(uint64(hypervisorConfig.NumVCPUs), 10)
+	annotations[hypannotations.DefaultMemory] = strconv.FormatUint(uint64(hypervisorConfig.MemorySize), 10)
+
+	req := &pb.CreateVMRequest{
+		Id:                   id,
+		Annotations:          annotations,
+		NetworkNamespacePath: network.NetworkID(),
+	}
+
+	res, err := s.client.CreateVM(context.Background(), req)
```
[the remark this answers is not in the document]
Comment (stevenhorsman) [reply src/runtime/virtcontainers/remote.go:89]: @yoheiueda - it was a long time ago, but do you remember the reason for this? Just to ensure this request doesn't have any timeout deadlines?

[the remark this answers is not in the document]
Comment (yoheiueda) [reply src/runtime/virtcontainers/remote.go:89]: @stevenhorsman I don't remember the reason why a new context is created here. I think it is safe to replace it with `ctx`.

[the remark this answers is not in the document]
Comment (katexochen) [reply src/runtime/virtcontainers/remote.go:89]: Same is also happening in the next function some lines further down.

**Comment by surajssd** [comment]
@stevenhorsman to test this IIUC. Build kata with these patches, build operator, and then deploy CAA with that operator? Does that sound ok?

**Review by wainersm — APPROVED** [review]
I didn't look at the content of the changes because a) they were reviewed already for CCv0, b) we have tested remote-hypervisor for a while now, c) those a cherry-picks

So passing CI this PR is ready to be merged.

**Comment by stevenhorsman** [comment]
> @stevenhorsman to test this IIUC. Build kata with these patches, build operator, and then deploy CAA with that operator? Does that sound ok?

Yeah - it's not super straight-forward as you need to update the CAA on depend on it. I did it previously and documented part of the process https://github.com/confidential-containers/confidential-containers/issues/134#issuecomment-1644417118 if that helps?

**Comment by wainersm** [comment]
> > a) they were reviewed already for CCv0
> 
> Please, don´t take this as something solid. We've merged **a lot** of content as part of the CCv0 assuming that it'd be reviewed when merged into the main branch.

ok, reverted my approval and will spend more time on reviewing it today.

File: src/runtime/hack/update-generated-hypervisor-proto.sh [hunk 2]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+
+set -o errexit -o pipefail -o nounset
+
+cd "$(dirname "${BASH_SOURCE[0]}")/.."
+
+protoc --gogottrpc_out=protocols/hypervisor \
```
Comment (wainersm) [inline src/runtime/hack/update-generated-hypervisor-proto.sh:9]: There is this effort to migrate out of gogo/protobuf (https://github.com/kata-containers/kata-containers/issues/7420) performed by @beraldoleal . I'd like to have Beraldo to look at this change to ensure we are not introducing more gogo protobuf that should be removed later.

Comment (stevenhorsman) [reply src/runtime/hack/update-generated-hypervisor-proto.sh:9]: I think Yohei followed the model of https://github.com/kata-containers/kata-containers/tree/main/src/runtime/protocols/cache, but if we want to move it somewhere else that's fine with me

Comment (beraldoleal) [reply src/runtime/hack/update-generated-hypervisor-proto.sh:9]: Hi @wainersm, thank you for ping me here.

While it is not adding any gogo specifics to the .proto file and its compiled artifacts, it is adding a new script that uses `--gogottrpc`. Currently, our other scripts are doing the same, but there is indeed an effort to drop it. Instead of `--gogottrpc` we should start using `--go-ttrpc_out`.

I'm not sure tho, at this point, if we can live with both at same time. So I'm fine with this as it is, and I can drop it when doing the removal in other places so we can drop it all at once.

Comment (beraldoleal) [reply src/runtime/hack/update-generated-hypervisor-proto.sh:9]: Now, about the location, I miss the `option go_package`, to understand where those should be placed after compiled.

All .proto files under `src/libs/protocols/protos` has that. And will need that soon on later.

Comment (beraldoleal) [reply src/runtime/hack/update-generated-hypervisor-proto.sh:9]: Thank you @stevenhorsman for adding the `go_package` option. Yes, if cache is there, I don't mind having another one. But we should start a discussion to make sure the current layout makes sense.

File: src/runtime/pkg/katautils/config.go [hunk 3]
```diff
@@ -1140,6 +1143,14 @@ func newDragonballHypervisorConfig(h hypervisor) (vc.HypervisorConfig, error) {
 	}, nil
 }
 
+func newRemoteHypervisorConfig(h hypervisor) (vc.HypervisorConfig, error) {
+
+	return vc.HypervisorConfig{
+		RemoteHypervisorSocket:  h.RemoteHypervisorSocket,
```
Comment (wainersm) [inline src/runtime/pkg/katautils/config.go:1149]: Small detail (a.k.a nit), other hypervisors define the default binary path in `src/runtime/pkg/katautils/config-settings.go` that is returned in case  of `h.path` is empty.  In our case, `/run/peerpod/hypervisor.sock` should be the default socket value, right?

See, for example, the following config initializer:
```
func newFirecrackerHypervisorConfig(h hypervisor) (vc.HypervisorConfig, error) {
	hypervisor, err := h.path()
	if err != nil {
		return vc.HypervisorConfig{}, err
	}
...
...
```
Also should we initialize the timeout with a default value below?

File: src/runtime/pkg/katautils/config.go [hunk 4]
```diff
@@ -1775,6 +1789,11 @@ func checkFactoryConfig(config oci.RuntimeConfig) error {
 // checkHypervisorConfig performs basic "sanity checks" on the hypervisor
 // config.
 func checkHypervisorConfig(config vc.HypervisorConfig) error {
+
+	if config.RemoteHypervisorSocket != "" {
```
Comment (wainersm) [inline src/runtime/pkg/katautils/config.go:1798]: ok, if remote hypervisor then does not perform the remaining checks. But at this point the socket should be created or not? I'm figuring out whether to check also the socket or not.

File: src/runtime/virtcontainers/hypervisor.go [hunk 5]
```diff
@@ -449,6 +457,15 @@ type HypervisorConfig struct {
 	// BlockiDeviceAIO specifies the I/O API to be used.
 	BlockDeviceAIO string
 
+	// The socket to connect to the remote hypervisor implementation on
```
Comment (surajssd) [inline src/runtime/virtcontainers/hypervisor.go:460]: Is this where the CAA exposes its socket?

File: src/runtime/virtcontainers/kata_agent.go [hunk 6]
```diff
@@ -744,37 +751,43 @@ func (k *kataAgent) startSandbox(ctx context.Context, sandbox *Sandbox) error {
...
-		return err
-	}
-	if _, err = k.updateRoutes(ctx, routes); err != nil {
-		return err
-	}
-	if err = k.addARPNeighbors(ctx, neighs); err != nil {
-		return err
+		// If a Policy has been specified, send it to the agent.
+		if len(sandbox.config.AgentConfig.Policy) > 0 {
+			if err := sandbox.agent.setPolicy(ctx, sandbox.config.AgentConfig.Policy); err != nil {
```
Comment (surajssd) [inline src/runtime/virtcontainers/kata_agent.go:766]: I am guessing we will do the policy work in subsequent PRs?

File: src/runtime/virtcontainers/remote.go [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+func (rh *remoteHypervisor) SaveVM() error {
+	return notImplemented("SaveVM")
+}
+
+func (rh *remoteHypervisor) ResumeVM(ctx context.Context) error {
+	return notImplemented("ResumeVM")
+}
+
+func (rh *remoteHypervisor) AddDevice(ctx context.Context, devInfo interface{}, devType DeviceType) error {
+	// TODO should we return notImplemented("AddDevice"), rather than nil and ignoring it?
```
Comment (surajssd) [inline src/runtime/virtcontainers/remote.go:202]: any specific reason to not log here as per the original commit?

File: src/runtime/pkg/katautils/config.go (hunk 4 above)
Comment (stevenhorsman) [reply src/runtime/pkg/katautils/config.go:1798]: I'll be honest I'm not sure - I was assuming that in the config.go then it was just validating the config is set-up properly and not that it had been actioned i.e. the actual socket exists - assuming that is what you are getting at?

File: src/runtime/virtcontainers/kata_agent.go (hunk 6 above)
Comment (stevenhorsman) [reply src/runtime/virtcontainers/kata_agent.go:766]: Yeah, it's not entirely clear to me the approach that we will use to get the policy passed over to the peer pod a the moment (and there was no policy code when this was originally written), so I've kept the simple approach for now

File: src/runtime/virtcontainers/remote.go (hunk 7 above)
Comment (stevenhorsman) [reply src/runtime/virtcontainers/remote.go:202]: Good question - I think if we return nil then we should log, but I'm not sure if it's better to return `notImplemented` rather than ignore it, hence I added the todo there. I'll leave the TODO so we have this discussion, but add the log statement back in so we don't forget it.

File: src/runtime/pkg/katautils/config.go (hunk 3 above)
Comment (stevenhorsman) [reply src/runtime/pkg/katautils/config.go]: I think I've understood the suggestion correctly, so added https://github.com/kata-containers/kata-containers/commit/5ee390461698bc1f3405b98b033d1e193a75ba19, so let me know if that works and I'll fix it up

File: src/runtime/virtcontainers/kata_agent.go (hunk 6 above)
Comment (surajssd) [reply src/runtime/virtcontainers/kata_agent.go:766]: Gotcha, so we take care of the policy work at once for peer pods!

File: src/runtime/virtcontainers/hypervisor.go (hunk 5 above)
Comment (surajssd) [reply src/runtime/virtcontainers/hypervisor.go:460]: Gotcha thanks Steve.

**Review by surajssd — APPROVED** [review]
Thanks @stevenhorsman

**Comment by ariel-adam** [comment]
@wainersm did you have a chance to review this and ack/nak?

**Comment by wainersm** [comment]
hi @gkurz ! The other day I was talking with @stevenhorsman about this PR, and we agreed that it would be good to have some "neutral review"  (i.e. from someone not involved directly on CoCo). I know that you fixed parts of QEMU implementation on virtcontainers and recently re-worked the qmp/hmp support, so I'm inclined to call you an expert on kata's hypervisors implementation :) That said, If you have free cycles && feels apt to then could you please help with reviews?

In a nutshell, this introduces the remote hypervisor type which establishes a very simple communication protocol over a local socket and where the kata runtime demands a non traditional hypervisor to create/start/stop VMs. The current only implementation of the remote hypervisor is peer pods (https://github.com/confidential-containers/cloud-api-adaptor/blob/main/docs/architecture.md#components).

File: src/runtime/virtcontainers/remote.go [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+	if tracer, ok := err.(interface{ StackTrace() errors.StackTrace }); ok {
+		for _, f := range tracer.StackTrace() {
+			logrus.Errorf("%+s:%d\n", f, f)
+		}
+	}
+
+	return err
+}
+
+func (rh *remoteHypervisor) PauseVM(ctx context.Context) error {
```
Comment (wainersm) [inline src/runtime/virtcontainers/remote.go:189]: Obviously the remote hypervisor is currently biased with peer pods implementation, so on my reviews here I've tried to think beyond peer pods. I was wondering if the hypervisor protocol should have more than just `CreateVMRequest`, `StartVMRequest` and `StopVMRequest` requests. It seems that `pause`, `save` and `resume`  should be defined on the protocol (rather than not implemented) and then "not implemented" on peer pods side.

File: src/runtime/virtcontainers/remote.go [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+	return "", "", notImplemented("GetVMConsole")
+}
+
+func (rh *remoteHypervisor) Disconnect(ctx context.Context) {
+	notImplemented("Disconnect")
+}
+
+func (rh *remoteHypervisor) Capabilities(ctx context.Context) types.Capabilities {
+	var caps types.Capabilities
+	caps.SetBlockDeviceHotplugSupport()
```
Comment (wainersm) [inline src/runtime/virtcontainers/remote.go:237]: Is it true for peer pods? i.e. does it support block device hotplug? Also wondering whether `Capabilities` should be part of hypervisor protocol or not.

File: src/runtime/virtcontainers/remote.go [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+func (rh *remoteHypervisor) HotplugRemoveDevice(ctx context.Context, devInfo interface{}, devType DeviceType) (interface{}, error) {
+	return nil, notImplemented("HotplugRemoveDevice")
+}
+
+func (rh *remoteHypervisor) ResizeMemory(ctx context.Context, memMB uint32, memoryBlockSizeMB uint32, probe bool) (uint32, MemoryDevice, error) {
+	return memMB, MemoryDevice{}, nil
+}
+
+func (rh *remoteHypervisor) GetTotalMemoryMB(ctx context.Context) uint32 {
+	return rh.config.MemorySize
```
Comment (wainersm) [inline src/runtime/virtcontainers/remote.go:220]: I don't know what's the semantic of `GetTotalMemoryMB` and the Hypervisor interface documentation does not provide any hint. It is called by sandbox's [`updateResources`](https://github.com/kata-containers/kata-containers/blob/main/src/runtime/virtcontainers/sandbox.go#L2156) to update resources (e.g. memory and cpu) dynamically.  The `updateResources` returns immediately if the sandbox configuration is set to [`static resources management`](https://github.com/kata-containers/kata-containers/blob/main/src/runtime/virtcontainers/sandbox.go#L169), this is the default configuration of remote hypervisor toml (https://github.com/kata-containers/kata-containers/pull/7046/commits/86f6762d827302895d1a45c08616a827d8f09f6c).

All that said, does it make sense GetTotalMemoryMB return anything other than zero maybe?

File: src/runtime/config/configuration-remote.toml.in [hunk 11]
```diff
@@ -242,6 +242,16 @@ disable_new_netns = true
...
 sandbox_cgroup_only=@DEFSANDBOXCGROUPONLY@
 
+# If enabled, the runtime will attempt to determine appropriate sandbox size (memory, CPU) before booting the virtual machine. In
+# this case, the runtime will not dynamically update the amount of memory and CPU in the virtual machine. This is generally helpful
+# when a hardware architecture or hypervisor solutions is utilized which does not support CPU and/or memory hotplug.
+# Compatibility for determining appropriate sandbox (VM) size:
+# - When running with pods, sandbox sizing information will only be available if using Kubernetes >= 1.23 and containerd >= 1.6. CRI-O
+#   does not yet support sandbox sizing annotations.
+# - When running single containers using a tool like ctr, container sizing information will be available.
+# Note: the remote hypervisor uses the peer pod config to determine the sandbox size, so requires this to be set to true
```
[the remark this answers is not in the document]
Comment (bpradipt) [reply src/runtime/config/configuration-remote.toml.in:252]: May be when there is an alternate implementation, during that time this can be changed ?
For now we only have peer-pods where dynamic resources handling is not possible. So imho, we should leave it as is and change it when there is a concrete alternative implementation. This is of course my personal view :-)

File: src/runtime/virtcontainers/remote.go (hunk 8 above)
Comment (bpradipt) [reply src/runtime/virtcontainers/remote.go:189]: Hmm, changing the hypervisor protocol will need some more discussions and also we need to study the impact on cloud-api-adaptor implementation. 
How about we deal with protocol extension as part of a new issue where we can document the reasons, requirements etc ?

**Comment by gkurz** [comment]
> hi @gkurz ! The other day I was talking with @stevenhorsman about this PR, and we agreed that it would be good to have some "neutral review" (i.e. from someone not involved directly on CoCo). I know that you fixed parts of QEMU implementation on virtcontainers and recently re-worked the qmp/hmp support, so I'm inclined to call you an expert on kata's hypervisors implementation :) That said, If you have free cycles && feels apt to then could you please help with reviews?
> 

Hey @wainersm and @stevenhorsman !

I'll find some cycles as this could simplify our downstream work. But you'll need to bear with me as I am totally CoCo-agnostic and I will ask silly questions ;-)

> In a nutshell, this introduces the remote hypervisor type which establishes a very simple communication protocol over a local socket and where the kata runtime demands a non traditional hypervisor to create/start/stop VMs. The current only implementation of the remote hypervisor is peer pods (https://github.com/confidential-containers/cloud-api-adaptor/blob/main/docs/architecture.md#components).

First silly question : this PR doesn't contains changes for the kata agent. Does this mean that peer pods don't need specific changes in the agent and I can use the agent from the main branch in the remote VM ?

File: src/runtime/config/configuration-remote.toml.in (hunk 11 above)
[the remark this answers is not in the document]
Comment (wainersm) [reply src/runtime/config/configuration-remote.toml.in:252]: @bpradipt a few minutes after I wrote that comment I changed my mind, and yes,  we better keep `configuration-remote` which defaults to the only implementation available: peer pods. So if in the future we end up with another implementation then we can start differentiating like `configuration-remote-caa`, `configuration-remote-foo`...etc

**Comment by wainersm** [comment]
> > hi @gkurz ! The other day I was talking with @stevenhorsman about this PR, and we agreed that it would be good to have some "neutral review" (i.e. from someone not involved directly on CoCo). I know that you fixed parts of QEMU implementation on virtcontainers and recently re-worked the qmp/hmp support, so I'm inclined to call you an expert on kata's hypervisors implementation :) That said, If you have free cycles && feels apt to then could you please help with reviews?
> 
> Hey @wainersm and @stevenhorsman !
> 
> I'll find some cycles as this could simplify our downstream work. But you'll need to bear with me as I am totally CoCo-agnostic and I will ask silly questions ;-)
>

I was unhappy of being the only one asking silly questions, so I appreciate you getting onboard :)

Seriously now, an CoCo-agnostic view is what we are looking for so don't worry.
  
> > In a nutshell, this introduces the remote hypervisor type which establishes a very simple communication protocol over a local socket and where the kata runtime demands a non traditional hypervisor to create/start/stop VMs. The current only implementation of the remote hypervisor is peer pods (https://github.com/confidential-containers/cloud-api-adaptor/blob/main/docs/architecture.md#components).
> 
> First silly question : this PR doesn't contain changes for the kata agent. Does this mean that peer pods don't need specific changes in the agent and I can use the agent from the main branch in the remote VM ?

Peer pods has the `agent-protocol-forward` app (https://github.com/confidential-containers/cloud-api-adaptor/blob/main/docs/architecture.md#agent-protocol-forwarder) which runs inside the VM to bridge the communication between kata-shim and `kata-agent`. This PR cover the introduction of the new remote hypervisor type and the infrastructure has that communication to happen. This is the bare-mininum required on kata's main branch for peer pods to work.

This PR **does not** cover the image offloading, which touches the kata agent code, and in simply terms allow to pull the container images directly inside the VM. On CCv0 branch we had this sole and simple approach of pulling the image from the VM, which required a fork of containerd to work. A while after, the containerd community implemented a mechanism to image offload using `snapshotters` and upon containerd's snapshotter specification we built a framework that allow different implementations of image offloading. Peer pods, however, still requires the old approach of pulling the image directly inside the VM.  Despite some tentative to track the progress of the image offload support in [this doc]( https://docs.google.com/document/d/14g38GK78wR32VsCume9KcMrBZLQMPrYMtBpdmcVpEZM/edit#bookmark=id.x6ctaejwqgy0), the reality is that the information is pulverized on pull requests, docs, issues, slack channels...etc. Thus yes, peer pods will need more than this PR to fully work from the `main` branch.

**Comment by gkurz** [comment]
> > 
> > Hey @wainersm and @stevenhorsman !
> > I'll find some cycles as this could simplify our downstream work. But you'll need to bear with me as I am totally CoCo-agnostic and I will ask silly questions ;-)
> 
> I was unhappy of being the only one asking silly questions, so I appreciate you getting onboard :)
> 

:rofl: 

> Seriously now, an CoCo-agnostic view is what we are looking for so don't worry.
> 
> > > In a nutshell, this introduces the remote hypervisor type which establishes a very simple communication protocol over a local socket and where the kata runtime demands a non traditional hypervisor to create/start/stop VMs. The current only implementation of the remote hypervisor is peer pods (https://github.com/confidential-containers/cloud-api-adaptor/blob/main/docs/architecture.md#components).
> > 
> > 
> > First silly question : this PR doesn't contain changes for the kata agent. Does this mean that peer pods don't need specific changes in the agent and I can use the agent from the main branch in the remote VM ?
> 
> Peer pods has the `agent-protocol-forward` app (https://github.com/confidential-containers/cloud-api-adaptor/blob/main/docs/architecture.md#agent-protocol-forwarder) which runs inside the VM to bridge the communication between kata-shim and `kata-agent`. This PR cover the introduction of the new remote hypervisor type and the infrastructure has that communication to happen. This is the bare-mininum required on kata's main branch for peer pods to work.
> 
> This PR **does not** cover the image offloading, which touches the kata agent code, and in simply terms allow to pull the container images directly inside the VM. On CCv0 branch we had this sole and simple approach of pulling the image from the VM, which required a fork of containerd to work. A while after, the containerd community implemented a mechanism to image offload using `snapshotters` and upon containerd's snapshotter specification we built a framework that allow different implementations of image offloading. Peer pods, however, still requires the old approach of pulling the image directly inside the VM. Despite some tentative to track the progress of the image offload support in [this doc](https://docs.google.com/document/d/14g38GK78wR32VsCume9KcMrBZLQMPrYMtBpdmcVpEZM/edit#bookmark=id.x6ctaejwqgy0), the reality is that the information is pulverized on pull requests, docs, issues, slack channels...etc. Thus yes, peer pods will need more than this PR to fully work from the `main` branch.

Thanks @wainersm ! This clarifies a lot, and drives me to the next question :smile:

Is this PR enough for the shim to properly drive an agent built against the CCv0-0.7.0 tag ?

**Comment by stevenhorsman** [comment]
> Thanks @wainersm ! This clarifies a lot, and drives me to the next question 😄
> 
> Is this PR enough for the shim to properly drive an agent built against the CCv0-0.7.0 tag ?

Not really unfortunately - it should be enough to drive the cloud-api-adaptor (as the implementation of a remote hypervisor via the `/run/peerpod/hypervisor.sock` and that can drive a remote kata-agent (though via the agent-protocol component as  the kata-agent expects to talk to a vsock)

https://github.com/confidential-containers/cloud-api-adaptor/blob/main/docs/architecture.md might help to explain some of the background and architecture, but please ask for clarification - there are no silly questions (but I can't promise that I won't give silly answers!)

File: src/runtime/protocols/hypervisor/hypervisor.proto [hunk 12]
```diff
@@ -0,0 +1,45 @@
+syntax = "proto3";
+
+option go_package = "github.com/kata-containers/kata-containers/src/runtime/protocols/hypervisor";
```
Comment (gkurz) [inline src/runtime/protocols/hypervisor/hypervisor.proto:6]: This change doesn't exist in the original patch. It is usually a good practice to mention manual modifications in cherry picked patches along with a justification.

Why is this line needed in main and not in CCv0 ?

Comment (stevenhorsman) [reply src/runtime/protocols/hypervisor/hypervisor.proto:6]: > This change doesn't exist in the original patch. It is usually a good practice to mention manual modifications in cherry picked patches along with a justification.

Hey Greg - that's a fair point there have been quite a lots of modifications to the original commits that I've cherry-picked in order to add unit tests, fix up todos and respond to review requests, so I should probably update most of them to say based on the original commit rather than a cherry pick

> Why is this line needed in main and not in CCv0 ?

This was requested by Beraldo above to line up with the best practice that the other `src/libs/protocols/protos` files use.

File: src/runtime/hack/update-generated-hypervisor-proto.sh (hunk 2 above)
Comment (beraldoleal) [reply src/runtime/hack/update-generated-hypervisor-proto.sh:9]: Hi @stevenhorsman since #8061 was merged already, I guess we have to migrate this away from gogo here now. Let me know if you need some help with it.

Comment (stevenhorsman) [reply src/runtime/hack/update-generated-hypervisor-proto.sh]: Hey Beraldo - thanks for reminding me about this. I've tried to do the migration, but if you have a chance to review what I've done and see if it works out then that would be appreciated!

**Review by beraldoleal — APPROVED** [review]
I revisted the protobuf/gogo removal and lgtm. Thanks @stevenhorsman

File: src/runtime/hack/update-generated-hypervisor-proto.sh (hunk 2 above)
Comment (wainersm) [reply src/runtime/hack/update-generated-hypervisor-proto.sh]: Closing this conversation as Beraldo review it already.

File: src/runtime/virtcontainers/kata_agent.go (hunk 6 above)
Comment (wainersm) [reply src/runtime/virtcontainers/kata_agent.go:767]: @surajssd @stevenhorsman does it need an extra issue to track the policy work for the remote hypervisor?

File: src/runtime/virtcontainers/remote.go (hunk 10 above)
Comment (stevenhorsman) [reply src/runtime/virtcontainers/remote.go:218]: Hey Wainer, sorry I think I missed this comment in the flurry of other reviews that day. Do you think that:
```
	//The remote hypervisor uses the peer pod config to determine the memory of the VM
	return 0
```
is a better statement there (assuming it doesn't break anything which might be tricky to find out at this point)?

Comment (stevenhorsman) [reply src/runtime/virtcontainers/remote.go:218]: Alternatively I could throw a `notImplemented` as we shouldn't get this far if static resource management is set?

File: src/runtime/pkg/katautils/config.go (hunk 3 above)
Comment (wainersm) [reply src/runtime/pkg/katautils/config.go]: hi @stevenhorsman , let's leave https://github.com/kata-containers/kata-containers/commit/5ee390461698bc1f3405b98b033d1e193a75ba19 for a follow up PR. Is it okay for you?

File: src/runtime/virtcontainers/remote.go (hunk 10 above)
Comment (stevenhorsman) [reply src/runtime/virtcontainers/remote.go]: How about https://github.com/kata-containers/kata-containers/pull/7046/commits/3d61038761d0db3c245dbc9bab6037dd2b246b7b ?

File: src/runtime/pkg/katautils/config.go (hunk 4 above)
Comment (wainersm) [reply src/runtime/pkg/katautils/config.go:1816]: `checkHypervisorConfig` does some checks like if the image is not empty, so I thought it wouldn't have sense to validate the remote socket existence. However, I am not sure if the socket file exist at this point.

Never mind, I think we can leave the code as is.

File: src/runtime/pkg/katautils/config.go (hunk 3 above)
Comment (stevenhorsman) [reply src/runtime/pkg/katautils/config.go]: I think I already fixup'd it after I didn't hear anything I'm afraid

File: src/runtime/virtcontainers/kata_agent.go (hunk 6 above)
Comment (stevenhorsman) [reply src/runtime/virtcontainers/kata_agent.go:767]: I've created https://github.com/kata-containers/kata-containers/issues/8443 for it so we don't forget. Thanks for the nudge

File: src/runtime/virtcontainers/remote.go (hunk 9 above)
Comment (wainersm) [reply src/runtime/virtcontainers/remote.go:237]: @stevenhorsman @bpradipt do you see any use of `Capabilities` in the protocol?

File: src/runtime/virtcontainers/remote.go (hunk 10 above)
Comment (stevenhorsman) [reply src/runtime/virtcontainers/remote.go]: I'm good at worrying about things! I think it's a far comment and you did a good investigation, so worth including

File: src/runtime/virtcontainers/remote.go (hunk 9 above)
Comment (bpradipt) [reply src/runtime/virtcontainers/remote.go:237]: @wainersm I think we'll need Capabilities in the protocol.  But not sure if BlockDeviceHotplugSupport is used somewhere.
@yoheiueda @huoqifeng do you recall if there is a need for SetBlockDeviceHotplugSupport ?

**Comment by bpradipt** [comment]
> @gkurz @surajssd @bpradipt @beraldoleal @katexochen @fidencio hi folks, you've reviewed this PR and @stevenhorsman has promptly addressed all comments. I didn't see any comment that would change the design of the remote hypervisor radically, instead only a few details of implementation has been noted. IMHO this is in good shape to land into main; but I would like to hear your final thoughts before I trigger the tests and begin the merge process. Thx!

@wainersm since some of the cherry-picked code was co-authored by me, I'm not in a position to review this PR :-)
However, personally I think this PR is ready to be merged.

File: src/runtime/virtcontainers/remote.go (hunk 9 above)
Comment (huoqifeng) [reply src/runtime/virtcontainers/remote.go:237]: I think to support CSI in remoteHyperVisor, BlockDeviceHotplugSupport is needed but I'll deffer it to @yoheiueda to confirm that.

Comment (yoheiueda) [reply src/runtime/virtcontainers/remote.go:237]: Yes. it has been introduced to support CSI in peer pods.

I think the capability was set when I was working on this problem with @genjuro214 https://github.com/kata-containers/kata-containers/pull/3697/commits/c3e6b660515046fef704e60d36efb05c10b5060b

**Review by wainersm — APPROVED** [review]
I'm not expert on this part of the project but I carefully reviewed the code by comparing with the other hypervisors implementations. Here we are following the pattern basically, and passing CI, so I'm confident it does not break the other hypervisors in any way.

@beraldoleal gave a very good review on the protocols side of the implementation. He has worked on the kata's protocols area for some time now, so I trust on his judgment.

**Review by jepio — APPROVED** [review]
Read through the code and nothing particularly evil jumps out at me.
